docs/job-writing-guide.md

```diff
@@ -0,0 +1,750 @@
+});
+post('/patients', dataValue('patients'));
+```
+
+To be clear, this is considered an anti-pattern and should not be used except in
+emergencies.
+
+## Reading state lazily
+
+A common problem in OpenFn coding is getting hold of the right state value at
+the right time.
+
+Consider this code:
+
+```js
+get('/some-data');
+post('/some-other-data', state.data);
+```
+
+What it's trying to do is call the GET method on some REST service, save the
+result to `state.data`, and then pass that `state.data` value into a post call
+to send it somewhere else.
+
+Can you see the problem?
+
+It happens all the time.
+
+Because of the way Javascript works, `state.data` will be evaluated before the
+`get()` request has finished. So the post will always receive a value of
+`undefined`.
+
+<!--
+Ok wise guy, how are we gonna explain this?
+It's something to do with factories, and closures, and synchronous execution.
+What's the simplest explanation though?
+-->
+<details>
+<summary>Okay, how <i>does</i> Javascript work?</summary>
+Javascript, like most languages, will evaluate synchronously, executing code line of code one at a time.
+
+Because each Operation is actually a factory function, it will execute instantly
+and return a function - but that function won't actually be executed yet.
+
+The returned function will have the original arguments in scope (trapped as
+closures), and be executed later against the latest state object.
+
+The example above will synchronously create two functions, which are added into
+an array by the compiler (see bellow), and then excuted by the runtime.
+
+So when your code executes, it's doing something like this:
+
+```
+const getFn = get('/some-data')
+const postFn = post('/some-data', state.data)
+
+return getFn(state).then((nextState) => postFn(nextState))
+```
+
+The point is that when the post operation is created, all we've done is _create_
+the `get` function - we haven't actually _run_ it. And so state.data is
+basically uninitialised.
+
+</details>
+
+What we actually need to do is defer the evaluation of `state.data` until the
+`post` operation actually runs. In other words, we work out the value of
+`state.data` at last possible moment.
+
+There are a few ways we can do that. Some jobs use `dataValue`, which is neat if
+a bit verbose (there are many examples in thie guide), and some operations
+support JSON path strings. The preferred way in modern OpenFn is to use an
+inline-function:
+
+```js
+get('/some-data');
+post('/some-other-data', state => state.data);
+```
+
+This passes a function into the `post` operator. When the `post()` call actually
+executes, the first thing it'll do is resolve any function into arguments into
+values. It does this by calling the function passing in the latest state, and
+using the return as the value.
+
+These lazy functions are incredibly powerful. Using them effectively is the key
+to writing good OpenFn jobs.
+
+## Iteration and each
```

Comment:
Given the content here - @josephjclark should we revisit [this page on the each() operation](https://docs.openfn.org/documentation/build/steps/each)?